hpx/parallel/util/loop.hpp

```diff
@@ -68,12 +71,194 @@ namespace hpx { namespace parallel { namespace util
+                operator++();
+                return tmp;
+            }
+            inline prefetching_iterator operator--(int)
+            {
+                prefetching_iterator tmp(*this);
+                operator--();
+                return tmp;
+            }
+            inline difference_type
+            operator-(const prefetching_iterator& rhs) const
+            {
+                return (idx-rhs.idx)/chunk_size;
+            }
+            inline prefetching_iterator
+            operator+(difference_type rhs) const
+            {
+                return prefetching_iterator((idx+(rhs*chunk_size)),
+                    (base+(rhs*chunk_size)),chunk_size,range_size,M_);
+            }
+            inline prefetching_iterator
+            operator-(difference_type rhs) const
+            {
+                return prefetching_iterator((idx-(rhs*chunk_size)),
+                    (base-(rhs*chunk_size)),chunk_size,range_size,M_);
+            }
+            friend inline prefetching_iterator
+            operator+(difference_type lhs,
+                const prefetching_iterator& rhs)
+            {
+                return rhs + lhs;
+            }
+            friend inline prefetching_iterator
+            operator-(difference_type lhs,
+                const prefetching_iterator& rhs)
+            {
+                return lhs - rhs;
+            }
+            inline bool operator==(const prefetching_iterator& rhs) const
+            {
+                return idx == rhs.idx;
+            }
+            inline bool operator!=(const prefetching_iterator& rhs) const
+            {
+                return idx != rhs.idx;
+            }
+            inline bool operator>(const prefetching_iterator& rhs) const
+            {
+                return idx > rhs.idx;
+            }
+            inline bool operator<(const prefetching_iterator& rhs) const
+            {
+                return idx < rhs.idx;
+            }
+            inline bool operator>=(const prefetching_iterator& rhs) const
+            {
+                return idx >= rhs.idx;
+            }
+            inline bool operator<=(const prefetching_iterator& rhs) const
+            {
+                return idx <= rhs.idx;
+            }
+            inline std::size_t operator*() const {return idx;}
+        };
+
+        ////////////////////////////////////////////////////////////////////
+
+        //Helper class to initialize prefetching_iterator
+        template<typename T>
+        struct prefetcher_context
+        {
+            std::vector<std::size_t> range;
+            std::vector<std::size_t>::iterator it_begin;
+            std::vector<std::size_t>::iterator it_end;
+            std::size_t chunk_size;
+            std::vector< T * > m;
+            std::size_t range_size;
+            explicit prefetcher_context (std::size_t begin, std::size_t end,
+                std::size_t p_factor, std::initializer_list< T * > &&l)
+            {
```

Comment:
Wouldn't it be better to expect the list of containers as a variadic template parameter list? I.e. something like:

    auto ctx = prefetcher(begin, end, factor, a, b, c);

where a, b, and c are the containers to prefetch.